### DISCUSSION TITLE
Make string failure messages more like Xunit for clarity?

### DISCUSSION THREAD

**Discussion by thomOrbelius** [fluentassertions/fluentassertions#2048] [Ideas]
Hi.

In my department we have been starting to use FluentAssertions. One of the selling point I had was the improved failure messages. A colleague of mine pointed out that the Xunit failure messages for strings is much more helpful (see example below) than the FluentAssertion equivalent. The example shows a case where we use a string with a newline separator.

Do any one else think that a message more similar to the one Xunit has is more helpful than the current? If so, should we update the message used?

```csharp
using FluentAssertions;
using Xunit;

namespace Sectra.Courier.RisAdapter.Tests.Integrations.SectraRis;

public class FluentComparison {
    private const string string1 = "String with \rnewline";
    private const string string2 = "String with \ra newline";

    [Fact]
    public void FluentAssert() {
        // Fails with:
        // newline" has a length of 20, differs near "new" (index 13).
        string1.Should().Be(string2);
    }

    [Fact]
    public void XUnitAssert() {
        // Fails with:
        // Assert.Equal() Failure
        //                         ↓ (pos 13)
        // Expected: String with \rnewline
        // Actual:   String with \ra newline
        //                         ↑ (pos 13)
        Assert.Equal(string1, string2);
    }
}
```

**Comment by jnyrup** [comment]
Hi Thom. 

This is definitely a case where we now lack behind the bare test frameworks and could improve. 
See also 
https://github.com/fluentassertions/fluentassertions/discussions/1941#discussioncomment-2921133

Have you noticed if there are other string APIs than `Be` that also lack behind?

**Reply by thomOrbelius** [reply]
I did some additional checks in the string API and compared to XUnit. I didn't find any other instances that felt too much of. I tested these:

```csharp
    [Fact]
    public void StartsWithXunit() => Assert.StartsWith("Anything else", "This");

    [Fact]
    public void StartsWithFluent() => "This".Should().StartWith("Anything else");

    [Fact]
    public void ContainsXunit() => Assert.Contains("Anything else", "This");

    [Fact]
    public void ContainsFluent() => "This".Should().Contain("Anything else");

    [Fact]
    public void MatchesXunit() => Assert.Matches("Anything else", "This");

    [Fact]
    public void MatchesFluent() => "This".Should().Match("Anything else");
```

And the typical output was

```csharp
//Fluent API
Expected string "This" to contain "Anything else".

// XUnit
Assert.Contains() Failure
Not found: Anything else
In value:  This
```

In my opinion I can go with both these messages.
That is, if any message is to be updated I think it is `Be` that needs attention.

**Comment by jnyrup** [comment]
I've created #2050 to track this work.
